[mpd] Initial support for stored playlists

chme opened the pull request:
The mpd protocol offers commands to create/edit/delete local m3u-playlists. This pr adds some support for it in forked-daapd.

The commands supported are:

- [x] Saving the queue to a playlist (or adding the queue to an existing playlist)
- [x] Adding a file/directory to a playlist (creates a new playlist if it does not exist)
- [x] Delete a playlist

Not supported commands are:
- [ ] Delete a song from a playlist
- [ ] Rearrange songs in a playlist
- [ ] Rename a playlist

The playlists are identified by their virtual path. The virtual path can be quite long, therefor i added a config option to define a default playlist folder. Playlists in this folder can be addressed simply by their name. 

I still need to do some more tests, but from my side this is good for a first review. The changes should not break existing functionality. This pr also includes the changes from #381.

ejurgensen commented:
Sorry for being so slow with this. FYI my plan is:
- squash bugs/stabilize for another ~2 weeks
- tag a new version
- merge this and other new features

chme commented:
No problem, i currently don't have much time either and did not finish my tests. So there is no need to hurry.

ejurgensen commented on src/library/filescanner.c:
```diff
@@ -1600,6 +1614,302 @@ scan_metadata(const char *path, struct media_file_info *mfi)
+  if (ret < 0)
+    return ret;
+
+  if (!fp)
```
Don't understand this check? If (!fp) then ret will be negative, so will this condition ever be true?

ejurgensen commented on src/library/filescanner.c:
```diff
@@ -1600,6 +1614,302 @@ scan_metadata(const char *path, struct media_file_info *mfi)
+static int
+playlist_add(const char *vp_playlist, const char *vp_item)
+{
+  FILE *fp = NULL;
```
Required to set to NULL? (see below comment)

ejurgensen commented on src/library/filescanner.c:
```diff
@@ -1600,6 +1614,302 @@ scan_metadata(const char *path, struct media_file_info *mfi)
+    return LIBRARY_ERROR;
+
+  ret = playlist_add_files(fp, pl_id, vp_item);
+  fclose(fp);
```
It might be nicer if fopen() and fclose() were in the same function? So that playlist_open() was instead used to check stuff and prepare a path for fopen().

ejurgensen commented on src/library/filescanner.c:
```diff
@@ -1600,6 +1614,302 @@ scan_metadata(const char *path, struct media_file_info *mfi)
+  if (ret < 0)
+    return ret;
+
+  if (!fp)
```
Same question as above

ejurgensen commented on src/listener.h:
```diff
@@ -16,6 +16,8 @@ enum listener_event_type
   LISTENER_OPTIONS   = (1 << 4),
   /* The library has been modified */
   LISTENER_DATABASE  = (1 << 5),
+  /* A stored playlist has ben modified (create, delete, add, rename) */
```
Typo

ejurgensen commented on forked-daapd.conf.in:
```diff
@@ -259,6 +259,11 @@ mpd {
+	# A directory in one of the library directories that will be used as the default
+	# playlist directory. forked-dapd creates new playlists in this directory if only
+	# a playlist name is provided by the mpd client (instead of a full virtual path)
+#	default_playlist_directory = "/srv/music/playlists"
```
Generally, if the setting is commented out, and there is a value, it means that this is the value forked-daapd will use unless otherwise configured. That is not the case here. My suggestion is to have an empty string here, and that the functionality is disabled by default (also serves as a security measure). Users who want this must then enter a directory.

ejurgensen commented on src/library/filescanner.c:
```diff
@@ -1600,6 +1614,302 @@ scan_metadata(const char *path, struct media_file_info *mfi)
+  *fp = fopen(pl_path, "a");
+  if (!(*fp))
+    {
+      DPRINTF(E_LOG, L_SCAN, "Error opening file '%s' for writing\n", pl_path);
```
Maybe echo the specific error with errno?

chme commented on src/library.h:
```diff
@@ -67,9 +67,29 @@ struct library_source
   int (*fullrescan)(void);
 
   /*
+   * Returns TRUE if the source supports items with the given virtual path
```
The intention was to get the source that is responsible to handle a given virtual path (find_source in library.c). But i dropped this function and used the same logic that is present in library_scan_media (iterating over all sources and returning after the first source reports it could handle the request).

chme commented on src/library/filescanner.c:
```diff
@@ -1600,6 +1614,302 @@ scan_metadata(const char *path, struct media_file_info *mfi)
+}
+
+static bool
+check_path_in_directories(const char *path)
```
I added a check, so that paths containing "/../" are invalid. realpath only works if the file already exists. And i guess symlinks are ok for directories in the library directory.

I hope this is good enough. But i am open to further suggestions. It should not be too critical as we always append ".m3u" to the given path. So only m3u files can be modified/created/deleted.

chme commented:
I think, i addressed all issues. Would be great, if you can take a second look at it. I rebased and squashed the commits and will do some more testing over the coming weekend.

ejurgensen commented:
Thanks a lot. Took it for a spin, and here are my remarks:
1. I would like writing to be disabled (i.e. impossible) if the config setting is not set.
2. I have forked-daapd running as root on my RPi (in my defence: forked-daapd is all it does), and if I make a symlink like e.g. "ln -s /etc/passwd /my/library/test.m3u" and then "mpc save file:/my/library/test" you can probably guess the result. What do you think about safeguarding against that?
3. If my library is /my/library, then I can do "mpc save file:/my/library", which will pass the check made by check_path_in_directories(). The playlist will be saved in /my/library.m3u, which actually isn't in the library dir.

Another thing: I have trouble with mpc and tab completion of the paths. Do you have it working?

zanzu commented:
Hi,

If I can interject, are you able to run forked-daapd with different userid?  Best practice when running daemons is to not run them as root UNLESS you absolutely have to.

Best Regards,
Charles

> On Aug 11, 2017, at 6:38 AM, ejurgensen <notifications@github.com> wrote:
> 
> Thanks a lot. Took it for a spin, and here are my remarks:
> 
> I would like writing to be disabled (i.e. impossible) if the config setting is not set.
> I have forked-daapd running as root on my RPi (in my defence: forked-daapd is all it does), and if I make a symlink like e.g. "ln -s /etc/passwd /my/library/test.m3u" and then "mpc save file:/my/library/test" you can probably guess the result. What do you think about safeguarding against that?
> If my library is /my/library, then I can do "mpc save file:/my/library", which will pass the check made by check_path_in_directories(). The playlist will be saved in /my/library.m3u, which actually isn't in the library dir.
> Another thing: I have trouble with mpc and tab completion of the paths. Do you have it working?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/ejurgensen/forked-daapd/pull/384#issuecomment-321680663>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ANnOz-H55saAYwRuJRqqPsD0ZpS24v1pks5sW3hjgaJpZM4NM_2r>.
>

chme commented:
> Another thing: I have trouble with mpc and tab completion of the paths. Do you have it working?

Only the server knows about the available paths. I don't think there is a way to get path completion in mpc (it would require making requests to the server). On my pc it does not work with mpd either.

> 1. I would like writing to be disabled (i.e. impossible) if the config setting is not set.

It feels a bit strange, that setting a path in the config will allow/disallow modifying playlists in totally different locations. 
I am not against disabling it by default, but would rather like to see another setting for it. Disabling it could be done in mpd.c or in library.c (or maybe filescanner.c). Depending on this, the config option should either be placed in the library section or the mpd section of the config file.

As for now only the mpd protocol supports it, i tending to keeping it in the mpd section. What do you prefer?

> 2. and if I make a symlink like e.g. "ln -s /etc/passwd /my/library/test.m3u" 

Linking a password file into a music directory, naming the link like a playlist and running a music app as root, that does not sound like a good idea :-)

I'll try to guard against this, not sure how yet. If you have a suggestion ...

> 3. If my library is /my/library, then I can do "mpc save file:/my/library", which will pass the check made by check_path_in_directories().

I guess only the directory part of the path should be checked against the configured library paths. I will fix this.

ejurgensen commented:
>Only the server knows about the available paths. I don't think there is a way to get path completion in mpc (it would require making requests to the server). On my pc it does not work with mpd either.

Yes, clearly it makes requests to the server :-) If I type "mpc add file:/" and then press tab, it seems to make a lsinfo request:
[2017-08-11 21:20:02] [DEBUG]      mpd: MPD message: lsinfo "file:"
This is also returned by forked-daapd to mpc, but it seems that mpc expects a slightly different result than what it gets, because on the command line it will then write: "mpc add file:file:/home/" (home is correct because the library is in /home/espen/Music).

This is on Ubuntu 16.04 with forked-daapd running on the same machine. It doesn't seem like mpc attempts tab completion when I specify another host (or -h localhost).

It also tries tab completion with "mpc save", though the end result for some reason is a bit different. Try it!

I know I am pushing it with the symlink and root, the point was just that a user with write access to the library can you achieve some privilege escalation, i.e. writing outside of the library with forked-daapd's privileges. So it is great if you can think about it.

Maybe use m_realpath() on the entire path, if it already exists, otherwise use it on the directory part of the path?

ejurgensen commented:
> As for now only the mpd protocol supports it, i tending to keeping it in the mpd section. What do you prefer?

I think your suggestion with a separate option sounds fine, and I would say the mpd section, since it is connected with the other option.

chme commented:
I added the new config option "allow_modifying_stored_playlists" to the mpd section and guarded against scenario 3.

I looked into using realpath, to check for symlinks, but it does only work for soft symlinks. I found no way to identify that a hard symlink is a symlink, even lstat gives false for S_ISLNK. forked-daapd scans the smylinks into its library, and scanning playlist files into the library and then not being able to modify them (after enabling it in the config file), also does not sound right to me.

If you insist, i will add the check for soft symlinks. But if you want to guard against symlinks, i think the better way would be to also disallow scanning them into the library (mpd has an option "follow_outside_symlinks", but it is enabled by default).

The created virtual path for symlinks (directories and files) does not look right. I look into this, but want to keep this outside of this pr.

ejurgensen commented:
Thanks, excellent. If it doesn't work for hardlinks I guess it is pointless to add the protection, so I merged without. If you come up with a solution it would be good, though.
